nvdaHelper/local/UIAEventLimiter/rateLimitedEventHandler.cpp

```diff
@@ -0,0 +1,243 @@
+{
+	LOG_DEBUG(L"RateLimitedUIAEventHandler::RateLimitedUIAEventHandler called");
+}
+
+// IUnknown methods
+ULONG STDMETHODCALLTYPE RateLimitedEventHandler::AddRef() {
+	auto refCount = InterlockedIncrement(&m_refCount);
+	LOG_DEBUG(L"AddRef: "<<refCount)
+	return refCount;
+}
+
+ULONG STDMETHODCALLTYPE RateLimitedEventHandler::Release() {
+	auto refCount = InterlockedDecrement(&m_refCount);
+	if (refCount == 0) {
+		delete this;
+	}
+	LOG_DEBUG(L"Release: "<<refCount)
+	return refCount;
+}
+
+HRESULT STDMETHODCALLTYPE RateLimitedEventHandler::QueryInterface(REFIID riid, void** ppInterface) {
+	if (riid == __uuidof(IUnknown)) {
+		*ppInterface = static_cast<IUIAutomationEventHandler*>(this);
+		AddRef();
+		return S_OK;
+	} else if (riid == __uuidof(IUIAutomationEventHandler)) {
+		*ppInterface = static_cast<IUIAutomationEventHandler*>(this);
+		AddRef();
+		return S_OK;
+	} else if (riid == __uuidof(IUIAutomationFocusChangedEventHandler)) {
+		*ppInterface = static_cast<IUIAutomationFocusChangedEventHandler*>(this);
+		AddRef();
+		return S_OK;
+	} else if (riid == __uuidof(IUIAutomationPropertyChangedEventHandler)) {
+		*ppInterface = static_cast<IUIAutomationPropertyChangedEventHandler*>(this);
+		AddRef();
+		return S_OK;
+	} else if (riid == __uuidof(IUIAutomationNotificationEventHandler)) {
+		*ppInterface = static_cast<IUIAutomationNotificationEventHandler*>(this);
+		AddRef();
+		return S_OK;
+	} else if (riid == __uuidof(IUIAutomationActiveTextPositionChangedEventHandler)) {
+		*ppInterface = static_cast<IUIAutomationActiveTextPositionChangedEventHandler*>(this);
+		AddRef();
+		return S_OK;
+	}
+	*ppInterface = nullptr;
+	return E_NOINTERFACE;
+}
+
+// IUIAutomationEventHandler method
+HRESULT STDMETHODCALLTYPE RateLimitedEventHandler::HandleAutomationEvent(IUIAutomationElement* sender, EVENTID eventID) {
+	LOG_DEBUG(L"RateLimitedUIAEventHandler::HandleAutomationEvent called");
+	LOG_DEBUG(L"Queuing automationEvent for eventID "<<eventID);
+	return queueEvent<AutomationEventRecord_t>(sender, eventID);
+}
+
+// IUIAutomationFocusEventHandler method
+HRESULT STDMETHODCALLTYPE RateLimitedEventHandler::HandleFocusChangedEvent(IUIAutomationElement* sender) {
+	LOG_DEBUG(L"RateLimitedUIAEventHandler::HandleFocusChangedEvent called");
+	return queueEvent<FocusChangedEventRecord_t>(sender);
+}
+
+// IUIAutomationPropertyChangedEventHandler method
+HRESULT STDMETHODCALLTYPE RateLimitedEventHandler::HandlePropertyChangedEvent(IUIAutomationElement* sender, PROPERTYID propertyID, VARIANT newValue) {
+	LOG_DEBUG(L"RateLimitedUIAEventHandler::HandlePropertyChangedEvent called");
+	return queueEvent<PropertyChangedEventRecord_t>(sender, propertyID, newValue);
+}
+
+// IUIAutomationNotificationEventHandler method
+HRESULT STDMETHODCALLTYPE RateLimitedEventHandler::HandleNotificationEvent(IUIAutomationElement* sender, NotificationKind notificationKind, NotificationProcessing notificationProcessing, BSTR displayString, BSTR activityID) {
```

Comment:
why not take a `NotificationEventRecord_t` here? same with the other `RateLimitedEventHandler`, why not take their equivalent records as parameters?

Reply:
These methods are declared by existing UI Automation event handler interfaces. E.g. IUIAutomationNotificationEventHandler. I can't change what parameters they take.